pages/clubs.jsx

```diff
@@ -82,63 +84,176 @@ var BottomCTA = React.createClass({
 });
 
 
-var ModalAddYourClub = React.createClass({
-  mixins: [React.addons.LinkedStateMixin, ModalManagerMixin],
+var ModalAddOrChangeYourClub = React.createClass({
+  mixins: [React.addons.LinkedStateMixin, ModalManagerMixin,
+           Router.Navigation, TeachAPIClientMixin],
   propTypes: {
-    onAddClub: React.PropTypes.func.isRequired
+    // If club is provided, then we're a 'change' dialog, otherwise
+    // we're an 'add' dialog.
+    club: React.PropTypes.object,
+    onSuccess: React.PropTypes.func.isRequired
   },
+  statics: {
+    teachAPIEvents: {
+      'username:change': 'handleUsernameChange',
+    }
+  },
+  STEP_AUTH: 1,
+  STEP_FORM: 2,
+  STEP_WAIT_FOR_NETWORK: 3,
+  STEP_SHOW_RESULT: 4,
   getInitialState: function() {
-    return {
+    var clubState = {
       name: '',
       website: '',
       description: '',
       location: ''
     };
+    if (this.props.club) {
+      _.extend(clubState, _.pick(this.props.club,
+        'name', 'website', 'description', 'location'
+      ));
+    }
+    return _.extend(clubState, {
+      step: this.getStepForAuthState(!!this.getTeachAPI().getUsername()),
+      result: null,
+      networkError: false
+    });
+  },
+  getStepForAuthState: function(isLoggedIn) {
+    return isLoggedIn ? this.STEP_FORM : this.STEP_AUTH;
+  },
+  handleUsernameChange: function(username) {
+    this.setState({step: this.getStepForAuthState(!!username)});
   },
   handleSubmit: function(e) {
-    e.preventDefault();
-    this.props.onAddClub(_.pick(this.state,
+    var teachAPI = this.getTeachAPI();
+    var clubState = _.pick(this.state,
       'name', 'website', 'description', 'location'
-    ), function(err) {
-      if (err) {
-        window.alert("Alas, an error occurred. Please try again later!");
-        console.log(err);
-      } else {
-        window.alert("Your club has been added!");
-        this.hideModal();
-      }
-    }.bind(this));
+    );
+
+    e.preventDefault();
+    this.setState({
+      step: this.STEP_WAIT_FOR_NETWORK,
+      networkError: false
+    });
+    if (this.props.club) {
+      clubState = _.extend({}, this.props.club, clubState);
+
+      // For now, let's always force re-geocoding.
```

Comment:
Lets file an issue to use Google location autocomplete and we always deal with this client-side rather than relying on the API to get the `lat` and `long` for us. IMO it doesn't seem to be reliable at all because we can get random location sometimes.